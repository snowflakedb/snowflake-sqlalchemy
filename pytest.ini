# pytest.ini
[pytest]
# There are majorly two issues with the sqlalchemy test suites if we want to run it with other snowflake specific tests:
# 1. If including the test_suite.py, only tests in test_suite.py will be collected, other dialect tests are not collected
# due to "from sqlalchemy.testing.plugin.pytestplugin import".
# 2. Putting test_suites.py and related conftest config under tests root dir would make pytest fail to collect cfg
# information such as requirements, and customized session db config does not work as expected (schema not set).
# Thus, we ignore the path for now, and give instructions to run the test suites in standalone directory.
addopts = --ignore=tests/sqlalchemy_test_suite
