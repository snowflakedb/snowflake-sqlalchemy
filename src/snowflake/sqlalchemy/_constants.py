#
# Copyright (c) 2012-2022 Snowflake Computing Inc. All rights reserved.
#
import pkg_resources

# parameters needed for usage tracking
PARAM_APPLICATION = "application"
PARAM_INTERNAL_APPLICATION_NAME = "internal_application_name"
PARAM_INTERNAL_APPLICATION_VERSION = "internal_application_version"

APPLICATION_NAME = "SnowflakeSQLAlchemy"
SNOWFLAKE_SQLALCHEMY_VERSION = pkg_resources.get_distribution(
    "snowflake-sqlalchemy"
).version
