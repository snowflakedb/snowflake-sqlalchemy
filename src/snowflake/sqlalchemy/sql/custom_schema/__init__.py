#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
#
from .dynamic_table import DynamicTable
from .hybrid_table import HybridTable

__all__ = ["DynamicTable", "HybridTable"]
