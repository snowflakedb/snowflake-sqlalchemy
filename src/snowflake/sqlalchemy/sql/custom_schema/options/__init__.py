#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
#

from .as_query import AsQuery
from .target_lag import TargetLag, TimeUnit
from .warehouse import Warehouse

__all__ = ["Warehouse", "AsQuery", "TargetLag", "TimeUnit"]
