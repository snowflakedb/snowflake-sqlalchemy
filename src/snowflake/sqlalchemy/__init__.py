#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
#

import sys

if sys.version_info < (3, 8):
    import importlib_metadata
else:
    import importlib.metadata as importlib_metadata

from sqlalchemy.types import (  # noqa
    BIGINT,
    BINARY,
    BOOLEAN,
    CHAR,
    DATE,
    DATETIME,
    DECIMAL,
    FLOAT,
    INT,
    INTEGER,
    REAL,
    SMALLINT,
    TIME,
    TIMESTAMP,
    VARCHAR,
)

from . import base, snowdialect  # noqa
from .custom_commands import (  # noqa
    AWSBucket,
    AzureContainer,
    CopyFormatter,
    CopyIntoStorage,
    CreateFileFormat,
    CreateStage,
    CSVFormatter,
    ExternalStage,
    JSONFormatter,
    MergeInto,
    PARQUETFormatter,
)
from .custom_types import (  # noqa
    ARRAY,
    BYTEINT,
    CHARACTER,
    DEC,
    DOUBLE,
    FIXED,
    GEOGRAPHY,
    GEOMETRY,
    MAP,
    NUMBER,
    OBJECT,
    STRING,
    TEXT,
    TIMESTAMP_LTZ,
    TIMESTAMP_NTZ,
    TIMESTAMP_TZ,
    TINYINT,
    VARBINARY,
    VARIANT,
)
from .sql.custom_schema import (  # noqa
    DynamicTable,
    HybridTable,
    IcebergTable,
    SnowflakeTable,
)
from .sql.custom_schema.options import (  # noqa
    AsQueryOption,
    ClusterByOption,
    IdentifierOption,
    KeywordOption,
    LiteralOption,
    SnowflakeKeyword,
    TableOptionKey,
    TargetLagOption,
    TimeUnit,
)
from .util import _url as URL  # noqa

base.dialect = dialect = snowdialect.dialect

__version__ = importlib_metadata.version("snowflake-sqlalchemy")

_custom_types = (
    "BIGINT",
    "BINARY",
    "BOOLEAN",
    "CHAR",
    "DATE",
    "DATETIME",
    "DECIMAL",
    "FLOAT",
    "INT",
    "INTEGER",
    "REAL",
    "SMALLINT",
    "TIME",
    "TIMESTAMP",
    "URL",
    "VARCHAR",
    "ARRAY",
    "BYTEINT",
    "CHARACTER",
    "DEC",
    "DOUBLE",
    "FIXED",
    "GEOGRAPHY",
    "GEOMETRY",
    "OBJECT",
    "NUMBER",
    "STRING",
    "TEXT",
    "TIMESTAMP_LTZ",
    "TIMESTAMP_TZ",
    "TIMESTAMP_NTZ",
    "TINYINT",
    "VARBINARY",
    "VARIANT",
    "MAP",
)

_custom_commands = (
    "MergeInto",
    "CSVFormatter",
    "JSONFormatter",
    "PARQUETFormatter",
    "CopyFormatter",
    "CopyIntoStorage",
    "AWSBucket",
    "AzureContainer",
    "ExternalStage",
    "CreateStage",
    "CreateFileFormat",
)

_custom_tables = ("HybridTable", "DynamicTable", "IcebergTable", "SnowflakeTable")

_custom_table_options = (
    "AsQueryOption",
    "TargetLagOption",
    "LiteralOption",
    "IdentifierOption",
    "KeywordOption",
    "ClusterByOption",
)

_enums = (
    "TimeUnit",
    "TableOptionKey",
    "SnowflakeKeyword",
)
__all__ = (
    *_custom_types,
    *_custom_commands,
    *_custom_tables,
    *_custom_table_options,
    *_enums,
)
