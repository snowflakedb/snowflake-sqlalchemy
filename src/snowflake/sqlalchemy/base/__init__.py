# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from sqlalchemy.schema import Table
from sqlalchemy.sql import functions

from ..functions import flatten
from .compiler.snowflake_compiler import SnowflakeCompiler
from .compiler.snowflake_ddl_compiler import SnowflakeDDLCompiler
from .compiler.snowflake_identifier_preparer import SnowflakeIdentifierPreparer
from .compiler.snowflake_type_compiler import SnowflakeTypeCompiler
from .constants import RESERVED_WORDS
from .snowflake_execution_context import SnowflakeExecutionContext
from .snowflake_orm_select_compile_state import SnowflakeORMSelectCompileState
from .snowflake_select_state import SnowflakeSelectState

# The __all__ list is used to maintain backward compatibility with previous versions of the package.
# After splitting a large file into a module, __all__ explicitly defines the public API, ensuring
# that existing imports from the original file still work as expected. This approach prevents
# breaking changes by controlling which functions, classes, and variables are exposed when
# the module is imported.
__all__ = [
    "SnowflakeDDLCompiler",
    "SnowflakeCompiler",
    "SnowflakeIdentifierPreparer",
    "SnowflakeTypeCompiler",
    "SnowflakeExecutionContext",
    "SnowflakeORMSelectCompileState",
    "SnowflakeSelectState",
    "RESERVED_WORDS",
]


construct_arguments = [(Table, {"clusterby": None})]

functions.register_function("flatten", flatten)
