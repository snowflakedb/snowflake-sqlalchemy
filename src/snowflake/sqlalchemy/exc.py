#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.

from typing import List

from sqlalchemy.exc import ArgumentError


class NoPrimaryKeyError(ArgumentError):
    def __init__(self, target: str):
        super().__init__(f"Table {target} required primary key.")


class UnsupportedPrimaryKeysAndForeignKeysError(ArgumentError):
    def __init__(self, target: str):
        super().__init__(f"Primary key and foreign keys are not supported in {target}.")


class RequiredParametersNotProvidedError(ArgumentError):
    def __init__(self, target: str, parameters: List[str]):
        super().__init__(
            f"{target} requires the following parameters: %s." % ", ".join(parameters)
        )


class UnexpectedTableOptionKeyError(ArgumentError):
    def __init__(self, expected: str, actual: str):
        super().__init__(f"Expected table option {expected} but got {actual}.")


class OptionKeyNotProvidedError(ArgumentError):
    def __init__(self, target: str):
        super().__init__(
            f"Expected option key in {target} option but got NoneType instead."
        )


class UnexpectedOptionParameterTypeError(ArgumentError):
    def __init__(self, parameter_name: str, target: str, types: List[str]):
        super().__init__(
            f"Parameter {parameter_name} of {target} requires to be one"
            f" of following types: {', '.join(types)}."
        )


class CustomOptionsAreOnlySupportedOnSnowflakeTables(ArgumentError):
    def __init__(self):
        super().__init__(
            "Identifier, Literal, TargetLag and other custom options are only supported on Snowflake tables."
        )


class UnexpectedOptionTypeError(ArgumentError):
    def __init__(self, options: List[str]):
        super().__init__(
            f"The following options are either unsupported or should be defined using a Snowflake table: {', '.join(options)}."
        )


class InvalidTableParameterTypeError(ArgumentError):
    def __init__(self, name: str, input_type: str, expected_types: List[str]):
        expected_types_str = "', '".join(expected_types)
        super().__init__(
            f"Invalid parameter type '{input_type}' provided for '{name}'. "
            f"Expected one of the following types: '{expected_types_str}'.\n"
        )


class MultipleErrors(ArgumentError):
    def __init__(self, errors):
        self.errors = errors

    def __str__(self):
        return "".join(str(e) for e in self.errors)


class StructuredTypeNotSupportedInTableColumnsError(ArgumentError):
    def __init__(self, table_type: str, table_name: str, column_name: str):
        super().__init__(
            f"Column '{column_name}' is of a structured type, which is only supported on Iceberg tables. "
            f"The table '{table_name}' is of type '{table_type}', not Iceberg."
        )
