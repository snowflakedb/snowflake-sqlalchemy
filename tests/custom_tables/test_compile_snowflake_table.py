#
# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
#
import pytest
from sqlalchemy import (
    Column,
    ForeignKeyConstraint,
    Integer,
    MetaData,
    String,
    select,
    text,
)
from sqlalchemy.exc import ArgumentError
from sqlalchemy.orm import declarative_base
from sqlalchemy.sql.ddl import CreateTable

from snowflake.sqlalchemy import SnowflakeTable
from snowflake.sqlalchemy.sql.custom_schema.options import (
    AsQueryOption,
    ClusterByOption,
)


def test_compile_snowflake_table(sql_compiler, snapshot):
    metadata = MetaData()
    table_name = "test_table_1"
    test_geometry = SnowflakeTable(
        table_name,
        metadata,
        Column("id", Integer),
        Column("geom", String),
        cluster_by=["id", text("id > 100")],
        as_query="SELECT * FROM table",
    )

    value = CreateTable(test_geometry)

    actual = sql_compiler(value)

    assert actual == snapshot


def test_compile_snowflake_table_with_explicit_options(sql_compiler, snapshot):
    metadata = MetaData()
    table_name = "test_table_2"
    test_geometry = SnowflakeTable(
        table_name,
        metadata,
        Column("id", Integer),
        Column("geom", String),
        cluster_by=ClusterByOption("id", text("id > 100")),
        as_query=AsQueryOption("SELECT * FROM table"),
    )

    value = CreateTable(test_geometry)

    actual = sql_compiler(value)

    assert actual == snapshot


def test_compile_snowflake_table_with_wrong_option_types(snapshot):
    metadata = MetaData()
    table_name = "test_snowflake_table"
    with pytest.raises(ArgumentError) as argument_error:
        SnowflakeTable(
            table_name,
            metadata,
            Column("id", Integer),
            Column("geom", String),
            as_query=ClusterByOption("id", text("id > 100")),
            cluster_by=AsQueryOption("SELECT * FROM table"),
        )

    assert str(argument_error.value) == snapshot


def test_compile_snowflake_table_with_primary_key(sql_compiler, snapshot):
    metadata = MetaData()
    table_name = "test_table_2"
    test_geometry = SnowflakeTable(
        table_name,
        metadata,
        Column("id", Integer, primary_key=True),
        Column("geom", String),
        cluster_by=ClusterByOption("id", text("id > 100")),
        as_query=AsQueryOption("SELECT * FROM table"),
    )

    value = CreateTable(test_geometry)

    actual = sql_compiler(value)

    assert actual == snapshot


def test_compile_snowflake_table_with_foreign_key(sql_compiler, snapshot):
    metadata = MetaData()

    SnowflakeTable(
        "table",
        metadata,
        Column("id", Integer, primary_key=True),
        Column("geom", String),
        ForeignKeyConstraint(["id"], ["table.id"]),
        cluster_by=ClusterByOption("id", text("id > 100")),
        as_query=AsQueryOption("SELECT * FROM table"),
    )

    table_name = "test_table_2"
    test_geometry = SnowflakeTable(
        table_name,
        metadata,
        Column("id", Integer, primary_key=True),
        Column("geom", String),
        ForeignKeyConstraint(["id"], ["table.id"]),
        cluster_by=ClusterByOption("id", text("id > 100")),
        as_query=AsQueryOption("SELECT * FROM table"),
    )

    value = CreateTable(test_geometry)

    actual = sql_compiler(value)

    assert actual == snapshot


def test_compile_snowflake_table_orm_with_str_keys(sql_compiler, snapshot):
    Base = declarative_base()

    class TestSnowflakeTableOrm(Base):
        __tablename__ = "test_snowflake_table_orm_2"

        @classmethod
        def __table_cls__(cls, name, metadata, *arg, **kw):
            return SnowflakeTable(name, metadata, *arg, **kw)

        __table_args__ = {
            "schema": "SCHEMA_DB",
            "cluster_by": ["id", text("id > 100")],
            "as_query": "SELECT * FROM table",
        }

        id = Column(Integer, primary_key=True)
        name = Column(String)

        def __repr__(self):
            return f"<TestSnowflakeTableOrm({self.name!r}, {self.fullname!r})>"

    value = CreateTable(TestSnowflakeTableOrm.__table__)

    actual = sql_compiler(value)

    assert actual == snapshot


def test_compile_snowflake_table_with_selectable(sql_compiler, snapshot):
    Base = declarative_base()

    test_table_1 = SnowflakeTable(
        "test_table_1",
        Base.metadata,
        Column("id", Integer, primary_key=True),
        Column("geom", String),
        ForeignKeyConstraint(["id"], ["table.id"]),
        cluster_by=ClusterByOption("id", text("id > 100")),
    )

    test_table_2 = SnowflakeTable(
        "snowflake_test_table_1",
        Base.metadata,
        as_query=select(test_table_1).where(test_table_1.c.id == 23),
    )

    value = CreateTable(test_table_2)

    actual = sql_compiler(value)

    assert actual == snapshot
