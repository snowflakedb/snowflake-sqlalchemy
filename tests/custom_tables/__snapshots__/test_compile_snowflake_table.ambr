# serializer version: 1
# name: test_compile_dynamic_table_orm_with_str_keys
  'CREATE TABLE "SCHEMA_DB".test_snowflake_table_orm_2 (\tid INTEGER NOT NULL AUTOINCREMENT, \tname VARCHAR, \tPRIMARY KEY (id))\tCLUSTER BY (id, id > 100)\tAS SELECT * FROM table'
# ---
# name: test_compile_dynamic_table_with_foreign_key
  'CREATE TABLE test_table_2 (\tid INTEGER NOT NULL, \tgeom VARCHAR, \tPRIMARY KEY (id), \tFOREIGN KEY(id) REFERENCES "table" (id))\tCLUSTER BY (id, id > 100)\tAS SELECT * FROM table'
# ---
# name: test_compile_dynamic_table_with_primary_key
  'CREATE TABLE test_table_2 (\tid INTEGER NOT NULL AUTOINCREMENT, \tgeom VARCHAR, \tPRIMARY KEY (id))\tCLUSTER BY (id, id > 100)\tAS SELECT * FROM table'
# ---
# name: test_compile_snowflake_table
  'CREATE TABLE test_table_1 (\tid INTEGER, \tgeom VARCHAR)\tCLUSTER BY (id, id > 100)\tAS SELECT * FROM table'
# ---
# name: test_compile_snowflake_table_orm_with_str_keys
  'CREATE TABLE "SCHEMA_DB".test_snowflake_table_orm_2 (\tid INTEGER NOT NULL AUTOINCREMENT, \tname VARCHAR, \tPRIMARY KEY (id))\tCLUSTER BY (id, id > 100)\tAS SELECT * FROM table'
# ---
# name: test_compile_snowflake_table_with_explicit_options
  'CREATE TABLE test_table_2 (\tid INTEGER, \tgeom VARCHAR)\tCLUSTER BY (id, id > 100)\tAS SELECT * FROM table'
# ---
# name: test_compile_snowflake_table_with_foreign_key
  'CREATE TABLE test_table_2 (\tid INTEGER NOT NULL, \tgeom VARCHAR, \tPRIMARY KEY (id), \tFOREIGN KEY(id) REFERENCES "table" (id))\tCLUSTER BY (id, id > 100)\tAS SELECT * FROM table'
# ---
# name: test_compile_snowflake_table_with_primary_key
  'CREATE TABLE test_table_2 (\tid INTEGER NOT NULL AUTOINCREMENT, \tgeom VARCHAR, \tPRIMARY KEY (id))\tCLUSTER BY (id, id > 100)\tAS SELECT * FROM table'
# ---
# name: test_compile_snowflake_table_with_selectable
  'CREATE TABLE snowflake_test_table_1 (\tid INTEGER, \tgeom VARCHAR)\tAS SELECT test_table_1.id, test_table_1.geom FROM test_table_1 WHERE test_table_1.id = 23'
# ---
# name: test_compile_snowflake_table_with_wrong_option_types
  '''
  Invalid parameter type 'AsQueryOption' provided for 'cluster by'. Expected one of the following types: 'ClusterByOption', 'list'.
  Invalid parameter type 'ClusterByOption' provided for 'as_query'. Expected one of the following types: 'AsQueryOption', 'str', 'Selectable'.
  
  '''
# ---
